doc/developer/design/20220214_create_cluster.md

```diff
@@ -0,0 +1,191 @@
+## Design
+
+### Cluster SQL syntax and semantics
+
+Four new SQL statements will be introduced with the following grammar to
+create, alter, drop, and list clusters, respectively:
+
+```
+create_cluster_stmt ::=
+  CREATE CLUSTER [IF NOT EXISTS] <name> [[WITH] <option> [, <option>...]]
+
+alter_cluster_stmt ::=
+  ALTER CLUSTER <name> [[SET] <option> [, <option>...]]
+
+drop_cluster_stmt ::=
+  DROP CLUSTER [IF EXISTS] <name> [CASCADE]
+
+show_cluster_stmt: SHOW CLUSTERS [{ LIKE 'pattern' | WHERE <expr> }]
+
+<option> ::=
+      VIRTUAL
+    | SIZE '<size>'
+
+<name> ::= <unqualified-identifier>
+```
+
+**All of these statements will be available in experimental mode only.**
+
+The name of a cluster is an unqualified identifier. Clusters are scoped to
+a catalog, and so do not belong in the usual namespace of databases and schemas.
+For a similarly scoped object in PostgreSQL, see [publications].
+
+The `CREATE CLUSTER` statement creates a new cluster. It returns an error if
+a cluster with the specified name already exists unless the `IF NOT EXISTS`
+clause is present.
+
+The `ALTER CLUSTER` statement changes properties of an existing cluster.
+
+The `DROP CLUSTER` statement drops an existing cluster. It returns an error
+if a cluster with the specified name does not exist unless the `IF EXIST`
+clause is present. Dropping a non-empty cluster (i.e., a cluster with at least
+one index) is not permitted unless `CASCADE` is present.
+
+THe `SHOW CLUSTERS` statement lists the clusters in the system, optionally
+filtered by the provided `LIKE` pattern or `WHERE` expression, which work
+analogously to the same clauses in the `SHOW DATABASES` statement.
+
+`CREATE CLUSTER` and `ALTER CLUSTER` allow specifying the same set of cluster
+options. Initially, there are only two options:
+
+  * `VIRTUAL` requests a virtual cluster. A virtual cluster may or may not
+    correspond to actual physical resources.
+
+    We suspect that the concept of a virtual cluster is primarily useful as a
+    way to exercise the cluster logic before we can allocate true clusters, and
+    it is likely that the `VIRTUAL` option is never stabilized. But they may
+    also prove to be a useful user-facing option, in which case we can consider
+    stabilization.
+
+  * `SIZE '<size>'` specifies a cluster of a fixed size. The size is an opaque
+    string from the perspective of `materialized`, but will be validated by the
+    cloud orchestration backend.
+
+    In the initial implementation, specifying the `SIZE` option will result in
+    an error during planning (i.e., not during parsing). The cloud team will be
+    along shortly to wire up this option to turn on actual physical clusters.
+
+Specifying both `VIRTUAL` and `SIZE` options in the same `CREATE CLUSTER` or
+`ALTER CLUSTER` is an error. Specifying the same option more than once is also
+an error.
+
+### Index SQL syntax and semantics
+
+The `CREATE INDEX` statement will be extended like so to allow specifying in
```

Comment:
Do `CREATE MATERIALIZED VIEW` and `CREATE MATERIALIZED SOURCE` also need to add `IN CLUSTER`?

Reply:
Good point. Updated.

Reply:
Do we need to add a syntax for tables (`CREATE MATERIALIZED TABLE`?) if tables are also going to have optional indexes? Are tables covered in other platform docs such that they don't need to be mentioned in this doc? I don't care too much, but feels worth a mention maybe.

Reply:
Internally, the likely thing to happen is that tables are created without indexes, which exist at the moment primarily due to the absence of persistent storage. You would be able to read them from any compute instance, just like any source, and if you would like to create an index on a compute instance you can using usual mechanism.

No opinion on what verbs get presented to the user, but wanted to spell out the likely internals (decoupling tables and indexes on them).

Reply:
That's a good point. Clusters expose that our current strategy of storing table data in its primary index is sketchy. @frankmcsherry had some thoughts about moving table data to be stored in the coordinator, and then the indexes in each cluster could be optional and function just like normal indexes.

Reply:
Yes, I'm on board with that. I think we might need to add a `MATERIALIZED` syntax to tables after that is implemented so they have analog features to sources and views. But that feature probably doesn't need to be listed in this document at all.